Do not allow checkout when item is not lockable or is already locked.

mauritsvanrees opened the pull request:
Fixes issue #135.

But tests are broken because we need to enable the `plone.locking` behavior on Folders in some tests do see the difference.  I try this, but this does not help.

Maybe we should adapt the tests so checkout is disallowed for Folders, and allowed for Documents.

mister-roboto commented:
@mauritsvanrees thanks for creating this Pull Request and helping to improve Plone!

TL;DR: Finish pushing changes, pass all other checks, then paste a comment:
```
@jenkins-plone-org please run jobs
```

To ensure that these changes do not break other parts of Plone, the Plone test suite matrix needs to pass, but it takes 30-60 min.  Other CI checks are usually much faster and the Plone Jenkins resources are limited, so when done pushing changes and all other checks pass either [start all Jenkins PR jobs yourself](https://jenkinsploneorg.readthedocs.io/en/latest/run-pull-request-jobs.html#run-a-pull-request-job), or simply add the comment above in this PR to start all the jobs automatically.

Happy hacking!

wesleybl commented on plone/app/iterate/browser/control.py:
```diff
@@ -39,6 +40,15 @@ class Control(BrowserView):
 
+    def _can_lock(self):
+        try:
+            if lock.isLocked(self.context):
+                return False
```
Is this correct? Does it return False if it is locked? It cannot be locked but it implements ILockable.

mauritsvanrees commented:
I have fixed the tests by introducing a `LockableFolder` portal_type.

After this, `tox -e dependencies` started complaining that `plone.app.contenttypes` was missing as dependency.  I fixed this by moving `profiles/testingdx` to `tests/profiles/testingdx.  And I removed the unused `profiles/test`.

Ready for review.  I want this in for the 6.1 release candidate.

mauritsvanrees commented:
@jenkins-plone-org please run jobs

wesleybl commented:
@mauritsvanrees looking at the failing tests, I see that they use `self.portal.docs`. For example:

https://github.com/plone/plone.app.iterate/blob/410c22409ed48d5dae501a103a2e7c05bd62a319/plone/app/iterate/tests/test_containers.py#L222

But this is a folder. And folder checkout support wasn't the strong point here. Maybe what the test author was wanting to do is test the permission on `self.portal.docs.doc1` which is a Document? Now with your implementation, it's fine to test on `self.portal.docs`, but I think we should test on `self.portal.docs.doc1` as well.

And it would also be good to test in a folder without ILockable.

wesleybl commented:
In fact, I was the one who added `.docs`, because previously only `self.portal` was used. I think the `checkout_allowed` method should be tested with the object we want to checkout, and not with its parent, correct?

wesleybl commented:
I think previously the `*_allowed_with_no_policy` tests were only for the portal. I should have created new tests for `self.portal.docs.doc1` instead of modifying the `self.portal` tests.

mauritsvanrees commented:
I see that `test_containers.py` and `test_iterate.py` have much in common.  I think the idea was that `test_containers.py` should test containers (folders) and `test_iterate.py` should test non-folderish items.
Okay, your commit makes sense then.

mauritsvanrees commented:
I am on it.

mauritsvanrees commented:
@wesleybl I have improved the tests further.  Can you have a look?

@jenkins-plone-org please run jobs

wesleybl commented on plone/app/iterate/tests/test_containers.py:
```diff
@@ -51,7 +55,7 @@ def setUp(self):
 
         # add a folder with two documents in it
-        self.portal.invokeFactory("Folder", "docs")
+        self.portal.invokeFactory("LockableFolder", "docs")
```
Since `docs` was already a normal folder and is used in other tests, I think it can continue to be a normal folder and the Lockable folder can be created in specific tests.

mauritsvanrees replied:
True.  I have fixed this, only now only use `LockableFolder` in one test.

wesleybl commented:
@jenkins-plone-org please run jobs
